docs/proposals/20231122-security-groups.md

```diff
@@ -0,0 +1,169 @@
+
+An operator will be able to decide how to manage the Security Groups for the Control Plane and Worker Nodes by using a new API. This new API will be used to create the Security Groups and the rules that are needed for the Control Plane and Worker Nodes.
+
+#### Data model changes
+
+`OpenStackCluster` will be changed and the proposed changes are purely additive and nonbreaking. No removals should be required to the data model. For `OpenStackCluster` we will add a new optional field in `spec.SecurityGroupsSpec` and then under it we'll have both `ControlPlaneSecurityGroupRules` and `WorkerSecurityGroupRules` which will represent the Security Groups for the Control Plane and Worker Nodes respectively.
+
+Example `OpenStackCluster` yaml:
+
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+kind: OpenStackCluster
+metadata:
+  name: cluster-sample
+  namespace: default
+  spec:
+    securityGroupsSpec:
+      controlPlaneSecurityGroupRules:
+        - direction: ingress
+          ethertype: IPv4
+          portRangeMin: 22
+          portRangeMax: 22
+          protocol: tcp
+        - direction: ingress
+          ethertype: IPv4
+          portRangeMin: 6443
+          portRangeMax: 6443
+          protocol: tcp
+        - direction: ingress
+          ethertype: IPv4
+          portRangeMin: 2379
+          portRangeMax: 2380
+          protocol: tcp
+        - direction: ingress
+          ethertype: IPv4
+          portRangeMin: 10250
+          portRangeMax: 10250
+          protocol: tcp
+      workerSecurityGroupRules:
+        - direction: ingress
+          ethertype: IPv4
+          portRangeMin: 22
+          portRangeMax: 22
+          protocol: tcp
+        - direction: ingress
+          ethertype: IPv4
+          portRangeMin: 10250
+          portRangeMax: 10250
+          protocol: tcp
+        - direction: ingress
+          ethertype: IPv4
+          portRangeMin: 80
+          portRangeMax: 80
+          protocol: tcp
+        - direction: ingress
+          ethertype: IPv4
+          portRangeMin: 443
+          portRangeMax: 443
+          protocol: tcp
+```
+
+We will replace the `managedSecurityGroups` field with `managedLegacySecurityGroups` which will be a boolean and will be used to determine if the legacy rules should be managed or not. This will be used to provide a migration path for users who were using the legacy rules provided by `OpenStackCluster.spec.managedSecurityGroups`. We'll add conversion logic to convert the old `managedSecurityGroups` field to the new `managedLegacySecurityGroups` field.
+
+Legacy rules and new rules will be managed in parallel, until we decide to remove the legacy rules. Which means that they can be used together, as they'll live in their own Security Groups.
```

Comment:
Do we really want/need that? I have mixed feelings

Reply:
>>>Legacy rules and new rules will be managed in parallel, until we decide to remove the legacy rules. Which means that they can be used together, as they'll live in their own Security Groups.

seems we don't , you provide a migration path and we should stick to newer way and drop older not good enough method to save effort and avoid maintain complexity

Reply:
After some discussion with Matt, i've reworked the whole KEP. Please have a look again.